TridiagSolver (distributed): "bulkerify" rank1 problem solution

albestro opened the pull request:
Following #860, this applies the same concepts to the distributed implementation.

Main changes:
- reduce micro-tasking
- rank1-solution is fully computed on MC even for GPU backend

TODO:
- [x] cleanup code/make it more readable
- [x] ~once current review is completed, merge forked #906~
- [x] #907

albestro committed:
minor changes

albestro committed:
first step: step1 laed4 + basic structure

albestro committed:
fix nthreads

albestro committed:
embed assembleDistEvalsVec in bulk

albestro committed:
embed step2 in bulk

albestro committed:
add basic version of step3 in bulk

albestro committed:
remove superfluous copies

albestro committed:
use blas::dot

albestro committed:
apply fix for k<=2 as done in last PR merged on master

albestro committed:
last step (normalization) reduce number of communications
by working by column-tile-wise instead of column-element-wise

albestro committed:
remove blocking mpi calls allowing yielding

albestro committed:
add missing barrier between multi to single threaded transition

albestro committed:
rename according to new guidelines + minor changes

albestro committed:
revert temporary change

albestro committed:
minor fixes

albestro committed:
start factoring out bcast and remove superfluous barriers

albestro committed:
remove sync_wait for bcast (it just wait at the end of bulk) + other changes

albestro committed:
minor changes

albestro commented:
cscs-ci run

albestro commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -959,135 +966,431 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+        // Note: this step is completely independent from the rest, but it is small and it is going
+        // to be dropped soon.
+        // Note: use last threads that in principle should have less work to do
+        // TODO if chunked this can be run by multiple workers
```
I would say that it will probably never get any improvement this block, because we are going to drop this as soon as we can improve the gemm. So I will remove this TODO 😉

albestro committed:
Update include/dlaf/eigensolver/tridiag_solver/merge.h

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

albestro committed:
add barrier busy wait + remove todo

albestro commented:
cscs-ci run

msimberg commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -961,135 +968,432 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+  }();
+
+  ex::start_detached(
+      ex::when_all(ex::just(std::make_shared<pika::barrier<>>(nthreads)),
```
I realized in the other PR that these can be `make_unique` instead of `make_shared`. The data isn't copied for each thread that is processing the `bulk`, but they all point to the same data, so there's no need for a `shared_ptr`. If you want you can update that here, but I will update other places in a separate PR.

albestro committed:
bug fix: race-condition over LAED4 workspace in case of multiple workers

fixed by re-using existing column workspaces, by adapting them per new
size requirements

albestro committed:
use make_unique instead of make_shared as suggested in #911

albestro committed:
minor change: reorder communicators (row first, col second)

albestro committed:
minor comment

albestro committed:
use "sub"-pipeline to force communications order on row communicator

albestro commented:
cscs-ci run

albestro committed:
remove unused capture

albestro commented:
cscs-ci run

albestro committed:
clang-format

rasolca commented:
cscs-ci run

msimberg reviewed (approved):
Approving despite comments. Most of the comments are just clarifying questions, and one comment is regarding a potential cleanup of the helper lambdas, but that is not a blocker for me.

albestro commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+  auto bcast_evals = [i_begin, i_end,
+                      dist](common::Pipeline<comm::Communicator>& row_comm_chain,
+                            const std::vector<matrix::Tile<T, Device::CPU>>& eval_tiles) {
+    using dlaf::comm::internal::sendBcast_o;
+    using dlaf::comm::internal::recvBcast_o;
+
+    const comm::Index2D this_rank = dist.rankIndex();
+
+    std::vector<ex::unique_any_sender<>> comms;
+    comms.reserve(to_sizet(i_end - i_begin));
+
+    for (SizeType i = i_begin; i < i_end; ++i) {
+      const comm::IndexT_MPI evecs_tile_rank = dist.rankGlobalTile<Coord::Col>(i);
+      auto& tile = eval_tiles[to_sizet(i - i_begin)];
+
+      if (evecs_tile_rank == this_rank.col())
+        comms.emplace_back(ex::when_all(row_comm_chain(), ex::just(std::cref(tile))) |
+                           transformMPI(sendBcast_o));
+      else
+        comms.emplace_back(ex::when_all(row_comm_chain(), ex::just(evecs_tile_rank, std::cref(tile))) |
+                           transformMPI(recvBcast_o));
+    }
+
+    return ex::ensure_started(ex::when_all_vector(std::move(comms)));
+  };
+
+  auto all_reduce_in_place_o = [](const dlaf::comm::Communicator& comm, MPI_Op reduce_op,
+                                  const auto& data, MPI_Request* req) {
+    auto msg = comm::make_message(data);
+    DLAF_MPI_CHECK_ERROR(MPI_Iallreduce(MPI_IN_PLACE, msg.data(), msg.count(), msg.mpi_type(), reduce_op,
+                                        comm, req));
+  };
+
+  // Note: at least two column of tiles per-worker, in the range [1, getTridiagRank1NWorkers()]
+  const std::size_t nthreads = [size = sz_loc_tiles.cols()]() {
+    const std::size_t min_workers = 1;
+    const std::size_t available_workers = getTridiagRank1NWorkers();
+    const std::size_t ideal_workers = util::ceilDiv(to_sizet(size), to_sizet(2));
+    return std::clamp(ideal_workers, min_workers, available_workers);
+  }();
```
Indeed, they were inside at some point, then I moved them out because I was thinking that probably some of them can be useful in general and can be added to the library itself. For the moment I wouldn't bother too much if you agree, because keeping them outside of the bulk reduces a bit the code cluttering in it (it is already too long and complex for my taste).

albestro commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+
+  const SizeType n = problemSize(i_begin, i_end, dist);
+
+  const SizeType m_subm_el_lc = [=]() {
+    const auto i_loc_being = ij_begin_lc.row();
+    const auto i_loc_end = ij_begin_lc.row() + sz_loc_tiles.rows();
+    return dist.localElementDistanceFromLocalTile<Coord::Row>(i_loc_being, i_loc_end);
+  }();
```
Yes. Moreover, I also thought about making a generic version over Row/Col (see also previous discussion about keeping these helpers outside of bulk) which could be made available via Distribution. But I didn't pursue this way for now since I have just one use-case for the moment (afaik).

albestro commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
-    // Fill ones for deflated Eigenvectors.
-    // Quick return if there is none.
-    if (n == k)
-      return;
-
-    const GlobalElementIndex origin{i_begin * dist.blockSize().rows(),
-                                    i_begin * dist.blockSize().cols()};
-    for (SizeType i = 0; i < n; ++i) {
-      const SizeType j = i2_ptr[i];
-      if (j >= k) {
-        const GlobalElementIndex i_g{i + origin.row(), j + origin.col()};
-        const GlobalTileIndex i_tile = dist.globalTileIndex(i_g);
-        if (dist.rankIndex() == dist.rankGlobalTile(i_tile)) {
-          const LocalTileIndex i_tile_l = dist.localTileIndex(i_tile);
-          const SizeType i_subm_evec_arr = i_tile_l.row() - idx_loc_begin.row() +
-                                           (i_tile_l.col() - idx_loc_begin.col()) * sz_loc_tiles.rows();
-          const TileElementIndex i = dist.tileElementIndex(i_g);
-          evec_tile_arr[to_sizet(i_subm_evec_arr)](i) = T{1};
+        // STEP 0b: Initialize workspaces (single-thread)
+        if (thread_idx == 0) {
+          // Note:
+          // - nthreads are used for both LAED4 and weight calculation (one per worker thread)
+          // - last one is used for reducing weights from all workers
+          ws_cols.reserve(nthreads + 1);
+
+          // Note:
+          // Considering that
+          // - LAED4 requires working on k elements
+          // - Weight computaiton requires working on m_subm_el_lc
+          //
+          // and they are needed at two steps that cannot happen in parallel, we opted for allocating the
+          // workspace with the highest requirement of memory, and reuse them for both steps.
+          const SizeType max_size = std::max(k, m_subm_el_lc);
+          for (std::size_t i = 0; i < nthreads; ++i)
+            ws_cols.emplace_back(max_size);
+          ws_cols.emplace_back(m_subm_el_lc);
+
+          ws_row = memory::MemoryView<T, Device::CPU>(n_subm_el_lc);
+          std::fill_n(ws_row(), n_subm_el_lc, 0);
         }
```
Yes, correct.

I didn't bother too much with trying to improve step0a because, as stated in the comment

```
// Note: this step is completely independent from the rest, but it is small and it is going
// to be dropped soon.
```

albestro commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+                             auto& ws_row) {
+        using dlaf::comm::internal::transformMPI;
+
+        common::Pipeline<comm::Communicator> row_comm_chain(row_comm_wrapper.get());
```
AFAIR this was not strictly needed, because we are launching them in order in a deterministic way...at least from our point of view. There might be details that I'm not  aware of: e.g. I submit tasks in order to a thread queue, but than it execute them in a different order).

In the end it looked safer and cleaner to use the pipeline.

msimberg replied:
I think this was indeed a potential issue
> I submit tasks in order to a thread queue, but than it execute them in a different order
so I think this is a good change. Thanks for explaining!

albestro committed:
rename all_reduce lambda as suggested

albestro commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+    return ex::ensure_started(ex::when_all_vector(std::move(comms)));
+  };
+
+  auto all_reduce_in_place_o = [](const dlaf::comm::Communicator& comm, MPI_Op reduce_op,
```
See https://github.com/eth-cscs/DLA-Future/pull/904/commits/ebe9beeaf187d98476ceacb675660449b7ec4081

albestro commented:
cscs-ci run

msimberg commented on include/dlaf/eigensolver/tridiag_solver/merge.h:
```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+  const SizeType n = problemSize(i_begin, i_end, dist);
+
+  const SizeType m_subm_el_lc = [=]() {
+    const auto i_loc_being = ij_begin_lc.row();
+    const auto i_loc_end = ij_begin_lc.row() + sz_loc_tiles.rows();
+    return dist.localElementDistanceFromLocalTile<Coord::Row>(i_loc_being, i_loc_end);
+  }();
+
+  const SizeType n_subm_el_lc = [=]() {
+    const auto i_loc_being = ij_begin_lc.col();
+    const auto i_loc_end = ij_begin_lc.col() + sz_loc_tiles.cols();
+    return dist.localElementDistanceFromLocalTile<Coord::Col>(i_loc_being, i_loc_end);
```
Sorry, one trivial fix that I forgot to comment about:
```suggestion
    const auto i_loc_begin = ij_begin_lc.row();
    const auto i_loc_end = ij_begin_lc.row() + sz_loc_tiles.rows();
    return dist.localElementDistanceFromLocalTile<Coord::Row>(i_loc_begin, i_loc_end);
  }();

  const SizeType n_subm_el_lc = [=]() {
    const auto i_loc_begin = ij_begin_lc.col();
    const auto i_loc_end = ij_begin_lc.col() + sz_loc_tiles.cols();
    return dist.localElementDistanceFromLocalTile<Coord::Col>(i_loc_begin, i_loc_end);
```

albestro committed:
Update include/dlaf/eigensolver/tridiag_solver/merge.h

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

rasolca commented:
cscs-ci run
